pageserver/src/tenant/timeline.rs

```diff
@@ -5691,6 +5689,14 @@ impl<'a> TimelineWriter<'a> {
             return OpenLayerAction::Open;
         };
 
+        if state.cached_last_freeze_at < self.tl.last_freeze_at.load() {
+            assert!(
+                state.open_layer.end_lsn.get().is_some(),
+                "our open_layer must be outdated"
+            );
```

Comment:
Is this assert correct? It doesn't fire in manual testing or with the test suite.

We always first write down the end_lsn then increment the `last_freeze_at`.

Reply:
The write_guard can be dragged down to here `LayerManager::try_freeze_whatever`, but the changes are quite scary looking, will do that on a follow-up.

Reply:
Until then, this assertion is valid, with that refactoring, we can remove this as it's dead code... Or make it a `#[cfg(feature = "testing")]` assertion, because then it will be clear that whenever we freeze we will clear this state out.

Reply:
We support a case where we don't have an open layer but we still we advance the `Timeline::last_freeze_at` to `last_record_lsn + 1`.

However it doesn't matter for the assertion, because we only have TimelineWriterState and get to this assertion when there is an inmemory layer which could be frozen.

Reply:
I created #7993 to separate the refactorings.